### ISSUE TITLE
Upgrade issues

### ISSUE THREAD

**Issue by distler** [fail2ban/fail2ban#1864]
Probably I just missed the "Upgrading" document, which delineated all the backward-incompatibilities in upgrading from 0.9.7 to 0.10.0. But, in the interest of easing the pain of others attempting the same upgrade (and, perhaps, encouraging the developers to include such a document more prominently in their distribution), I will enumerate the ones I encountered

1. Switching from `.egg` to `.egg-info` format (without mentioning the switch to the user), means that the installation *won't* work out of the box. The user has to *manually* edit the `site-packages/easy-install.pth` to remove the `./fail2ban-0.9.x.egg` line.
2. In filters, the locution `(?P=host)` no longer works. Instead, you need to write `((?P=ip4)|(?P=ip6))`.
3. In jails, it is no longer sufficient to write
    ```
    action = pf
   ```
   You need to specify a `port` (or port range) and `name`, e.g
   ```
   action   = pf[port={80 443}, name=http]
   ```
4. Finally (and most seriously), there's  Issue #1863 .

Presumably, there are other issues that I haven't encountered yet. But, at least, now I have both my `pf` firewall and fail2ban running again.

**Comment by sebres** [comment]
> But, in the interest of easing the pain of others attempting the same upgrade (and, perhaps, encouraging the developers to include such a document more prominently in their distribution)...

- Please take a look in [release of 0.10](https://github.com/fail2ban/fail2ban/releases/tag/0.10.0).
What do you see **in bold** there?
- The release is normally made for maintainers of some distribution (that can and should make own review and adjust its customization rules) or for experienced users (that do the upgrade on your own risk).
All other could wait until it will be released in distribution repository.

> Switching from .egg to .egg-info format (without mentioning the switch to the user), means that the installation won't work out of the box. The user has to manually edit the site-packages/easy-install.pth to remove the ./fail2ban-0.9.x.egg line.

- What did you installed exactly? We have made our issue-template not for nothing! More info would be very desirable...
- I don't know, that we have something about the handling with "egg" or "egg-info" in our repository.
- If you mean this belongs to this repo, can you show us where exactly resp. which commit do that.

> In filters, the locution (?P=host) no longer works. Instead, you need to write ((?P=ip4)|(?P=ip6))

- This going to ipv6 possibility, more complex rule-set to define parameter groups (dependency to `usedns` resp. `raw`, ipv6-ability etc.).
- Where it is documentated (as possibility to use in this case)? You would find `<HOST>` but not `(?P<host>)`.
- You can use you own groups (like `_cond_ip_` below):
```bash
fail2ban-regex "1500000000 failure from 192.0.2.4: bad host 192.0.2.4" "^\s*failure from (?P<_cond_ip_><HOST>): bad host (?P=_cond_ip_)$"
```

> In jails, it is no longer sufficient to write `action = pf`. 

I doubt this going to the 0.10th release. Parameter `banaction` as well as the interpolation logic were introduced several years ago (still by 0.9.0 or even earlier).

**Comment by distler** [comment]
>Please take a look in release of 0.10.
What do you see in bold there?

That there would be backwards-incompatible changes. A list  of such would be helpful.

>* What did you installed exactly? We have made our issue-template not for nothing! More info would be very desirable...
>* I don't know, that we have something about the handling with "egg" or "egg-info" in our repository.

In previous (0.9.x) versions,
```
  sudo python setup.py install
```
would create a `site-path/fail2ban-0.9.x-py2.x.egg` directory and update `site-packages/easy-install.pth` appropriately. The upgrade proceeds without a hiccup.

In the current version, 
```
  sudo python setup.py install
```
creates a `site-path/fail2ban-0.10.0-py2.x.egg-info` and a `site-path/fail2ban` directory.

Personally, I think `.egg` is a better format for seamless upgrading, but if y'all prefer `.egg-info`, that's fine.

The problem arises in *switching* from the one format to the other. The naive user (and I am certainly one) who had a previous version of fail2ban installed, will not know that manually editing `site-packages/easy-install.pth` is *required* in order for Python to pick up the new version when starting fail2ban.

Instead, running `fail2ban-client` just yielded an inscrutable Python error.

> * Where it is documentated (as possibility to use in this case)? You would find <HOST> but not (?P<host>).

`<HOST>` previously populated the group `host`, which could then be referred back to, using `(?P=host)`. Now it no longer populates that group, so previously-working filters now break. That's fine. You warned that there would be backwards-incompatible changes. I just want to document them ...

**Comment by sebres** [comment]
> ... backwards-incompatible changes. A list of such would be helpful.

You may inspect a ChangeLog to find all things incompatible to you.

We've no chances to catch all imaginable incompatibilities (of all possible users configs, a lot of platforms and distributions, maintainer customization's, etc.).

> The problem arises in switching from the one format to the other. 

As I already said "What did you installed exactly?" and "I don't know about changes  egg -\> egg-info in our repo".

**Comment by distler** [comment]
>You may inspect a ChangeLog to find all things incompatible to you.

I did.

But nowhere there is it mentioned (to pick one example) that `<HOST>` no longer populates the `host` group. To find that out, I actually had to stare at the source code.

> As I already said "What did you installed exactly?"

1. I had a (actually, several) previous 0.9.x version(s) already installed.
2. I did the standard
    ```
    sudo python setup.py install
   ```
   to install 0.10.0. That's it; I did nothing different from installing previous (0.9.x) versions.
 
> and "I don't know about changes egg -> egg-info in our repo".

The choice of packaging format (`.egg` vs. `.egg-info`) is up to the package maintainer, not the end-user.

I'm surprised that you're surprised that your own packaging format has changed from 0.9.x to 0.10.x. But you can easily check for yourself what sort of package is installed.

**Comment by sebres** [comment]
> I'm surprised that you're surprised that your own packaging format has changed from 0.9.x to 0.10.x. 
> I had previous 0.9.x version(s) - `debian/0.9.7-1`.
> Install 0.10.0 (downloaded `0.10.0`).

So you had a **debian** release and hereafter installed a **common** release (for all distributions, without debian-related customization)... 
Why I'm not surprised that it stutters? :)

In our repo we creates 2 releases: common one, e. g. 0.9.7 and debian-related debian/0.9.7-1 (@yarikoptic creates this for debian/neurodebian).
For 0.10 currently was created only 0.10.0 common release.

**Comment by distler** [comment]
>For 0.10 currently was created only 0.10.0 common release.

And so this issue will arise for anyone who previously installed a -debian- release.

**Comment by sebres** [comment]
> And so this issue will arise for anyone who previously installed a -debian- release.

You cannot be serious with this statement...

**Comment by sebres** [comment]
> 1. Switching from .egg to .egg-info format 
> 2. In filters, the conditional match `(?P=host)` no longer works.
> 3. In jails, it is no longer sufficient to write `action = pf`.
> 4. Finally (and most seriously), there's Issue #1863 .

So I'll summarize here:
1. Because of usage of common release as upgrade to debian release.
2. Info about capture-group `(?P=host)` together with another known incompatibilities added to [ChangeLog](/fail2ban/fail2ban/blob/0.10/ChangeLog) as "Incompatibility list" section. 
3. IMHO a pure configuration error (see discussion above about the `banaction`), but to avoid later such kinds of errors I've additionally fixed it in 08646bc3399f62d6a7eed87579f72614fd3e9355.
4. Fixed in PR #1866.

Thanks for reporting this!
Thus I'm closing.
